fuse_core/src/time.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2020, Brett Downing
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+
+/*
+ * This file provides a sane configuration for a clock to manage order of events.
+ * This file exists to help with a migration from fuse_core::TimeStamp to rclcpp::time
+ * Priority is given to C++ std::chrono
+ * emphasis will be placed on interoperability with fuse_core::TimeStamp and rclcpp::time
+ *
+ * This will likely be extended to distinguish between:
+ *    times of events being optimised
+ *    events generated by the running of the optimiser (log throttling, callbacks etc)
+ *
+ */
+
+#include <fuse_core/time.h>
+#include <rclcpp/utilities.hpp>
+
+using chrono_ns_time_point_t =
+  std::chrono::time_point<std::chrono::system_clock, std::chrono::nanoseconds>;
+
+namespace fuse_core
+{
+TimeStamp::TimeStamp() :
+  rclcpp::Time(static_cast<int64_t>(0), RCL_CLOCK_UNINITIALIZED)
```

Comment:
I thought time zero was explicitly removed from the ROS1 API?
It certainly makes little sense for ROS2; nodes have no fixed datum and there's no guarantee they won't cross zero at runtime. (I'm dealing with related issues in a GStreamer package)

I only built time zero into fuse timestamp to make it easier to remove it from the fuse_publisher logic, I intended to delete it during cleanup
`rclcpp::Time::Zero` could do more harm than good if it gets up-streamed.

Reply:
> It certainly makes little sense for ROS2; nodes have no fixed datum and there's no guarantee they won't cross zero at runtime. (I'm dealing with related issues in a GStreamer package)

I don't think crossing zero is a concern for ROS 2. At the lowest level [`rcutils` forbids negative time from the system and steady clocks](https://github.com/ros2/rcutils/blob/98d624a455e52c8bdd150c380ca302cf6a8a750c/src/time_unix.c#L69-L72). The [`rclcpp::Time` class tries to forbid negative time too](https://github.com/ros2/rclcpp/blob/edbfe1404b24d0bc85ff88e8ff1f006670788e46/rclcpp/src/rclcpp/time.cpp#L52-L54), but I see at least one bug there that I'll go fix.

Reply:
I didn't upstream this specifically because I didn't know what clock type to use as default. Well that and I don't think there should be a clock type that is used by default.

So making a Zero static class member with a particular clock type would bait a lot of users into getting runtime errors if they, for instance, used it for comparison.